core/primitives-core/src/types.rs

```diff
@@ -45,3 +45,70 @@ pub type ReceiptIndex = usize;
 pub type PromiseId = Vec<ReceiptIndex>;
 
 pub type ProtocolVersion = u32;
+
+/// The shard identifier. It may be a arbitrary number - it does not need to be
+/// a number in the range 0..NUM_SHARDS. The shard ids do not need to be
+/// sequential or contiguous.
+///
+/// The shard id is wrapped in a newtype to prevent the old pattern of using
+/// indices in range 0..NUM_SHARDS and casting to ShardId. Once the transition
+/// if fully complete it potentially may be simplified to a regular type alias.
+///
+/// TODO get rid of serde
+#[derive(
+    arbitrary::Arbitrary,
+    borsh::BorshSerialize,
+    borsh::BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Hash,
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct ShardId(u64);
+
+impl ShardId {
+    /// Create a new shard id. Please note that this function should not be used
+    /// directly. Instead the ShardId should be obtained from the shard_layout.
+    pub fn new(id: u64) -> Self {
+        Self(id)
+    }
+
+    /// Get the numerical value of the shard id. This should not be used as an
+    /// index into an array, as the shard id may be any arbitrary number.
+    pub fn get(self) -> u64 {
```

Comment:
Why don't we keep `shard_id.into()` as primary conversion method instead of `shard_id.get()` while we are sorting this mess? That looks more natural to me and the calls to `.into()` are always explicit in rust

Reply:
Similarly, we could use `ShardId::from(id)` instead of `ShardId::new(id)`

Reply:
+1 for `into`. see my similar comment about operations involving u64 or other integers as well.